### PR TITLE
Prefer HashNoOp for Parallel Append [#130041367]

### PR DESCRIPTION
**Pull request by dhanashreek89**
This is done by changing cost model to reduce the Motion cost when distribution spec
is CDistributionSpecHashedNoOp.

There are three distribution specs requested by `CPhysicalParallelUnionAll`:
- `CDistributionSpecHashedNoOp`: To produce best performance plan to align with base table
  distribution.
- `CDistributionSpecStrictHashed`: To produce parallel append plan by shuffling tuples using
  all distributable columns in the project list of UNION ALL.
- `CDistributionSpecStrictRandom`: To produce parallel append plan by shuffling all tuples
  using random distribution.

Most updates to the minidump files are reflecting the costing change, and also the plan id
change due to costing change.

We also output DistributionSpec types under TF 101010 to clearly indicate type of spec requested.

### CODE REVIEW

File: libgpopt/include/gpopt/base/CDistributionSpecHashedNoOp.h [hunk 1]
```diff
@@ -21,6 +21,11 @@ namespace gpopt
 
 			virtual BOOL FMatch(const CDistributionSpec *pds) const;
 
+			virtual const CHAR *SzId() const
+			{
+				return "HASH NOOP";
```
Comment (vraghavan78) [inline libgpopt/include/gpopt/base/CDistributionSpecHashedNoOp.h]: NO-OP

File: libgpdbcost/src/CCostModelGPDB.cpp [hunk 2]
```diff
@@ -1168,6 +1169,16 @@ CCostModelGPDB::CostMotion
 			+
 			pci->DRows() * pci->DWidth() * dRedistributeRecvCostUnit
 		));
+
+		// Adjust cost based on distribution spec
+		// to make sure parallel append picked the most optimal plan
```
Comment (vraghavan78) [inline libgpdbcost/src/CCostModelGPDB.cpp]: update comment: adjust the cost of no-op strict distribution to correctly reflect that no effort is needed for this.

File: libgpdbcost/src/CCostModelGPDB.cpp [hunk 3]
```diff
@@ -1168,6 +1169,16 @@ CCostModelGPDB::CostMotion
...
+		CPhysicalMotion *pMotion = CPhysicalMotion::PopConvert(exprhdl.Pop());
+		CDistributionSpec *pds = pMotion->Pds();
+		if (CDistributionSpec::EdtHashedNoOp == pds->Edt())
+		{
+			// promote the plan with redistribution on same distributed columns of base table for parallel append
+			costLocal = costLocal * CCost(0.9);
```
Comment (vraghavan78) [inline libgpdbcost/src/CCostModelGPDB.cpp]: Why not do something targeted for NoOp like we do for broadcast?

File: data/dxl/minidump/CPhysicalParallelUnionAllTest/TwoHashedTables.mdp [hunk 4]
```diff
@@ -225,110 +225,110 @@ EXPLAIN SELECT * FROM foo UNION ALL SELECT * FROM bar;
                 </dxl:LogicalGet>
             </dxl:UnionAll>
         </dxl:Query>
-        <dxl:Plan Id="0" SpaceSize="3">
```
Comment (vraghavan78) [inline data/dxl/minidump/CPhysicalParallelUnionAllTest/TwoHashedTables.mdp]: we reduced the search space. Can you see what was the plan we lost? It is a small space to RCA.

File: data/dxl/minidump/CPhysicalParallelUnionAllTest/ParallelUnion-Insert.mdp [hunk 5]
```diff
@@ -6,7 +6,7 @@ INSERT INTO t VALUES (11),(12),(13);
 <dxl:DXLMessage xmlns:dxl="http://greenplum.com/dxl/2010/12/">
   <dxl:Thread Id="0">
     <dxl:OptimizerConfig>
-      <dxl:EnumeratorConfig Id="2" PlanSamples="0" CostThreshold="0"/>
```
Comment (vraghavan78) [inline data/dxl/minidump/CPhysicalParallelUnionAllTest/ParallelUnion-Insert.mdp]: nice!

File: data/dxl/minidump/CPhysicalParallelUnionAllTest/ParallelUnion-ConstTable.mdp [hunk 6]
```diff
@@ -6,7 +6,7 @@ SELECT 1 UNION ALL SELECT * FROM t;
 <dxl:DXLMessage xmlns:dxl="http://greenplum.com/dxl/2010/12/">
   <dxl:Thread Id="0">
     <dxl:OptimizerConfig>
-      <dxl:EnumeratorConfig Id="2" PlanSamples="0" CostThreshold="0"/>
```
Comment (vraghavan78) [inline data/dxl/minidump/CPhysicalParallelUnionAllTest/ParallelUnion-ConstTable.mdp]: why 3?

File: data/dxl/minidump/CPhysicalParallelUnionAllTest/ParallelUnion-ConstTable.mdp [hunk 7]
```diff
@@ -211,7 +211,7 @@ SELECT 1 UNION ALL SELECT * FROM t;
         </dxl:LogicalGet>
       </dxl:UnionAll>
     </dxl:Query>
-    <dxl:Plan Id="2" SpaceSize="3">
```
Comment (vraghavan78) [inline data/dxl/minidump/CPhysicalParallelUnionAllTest/ParallelUnion-ConstTable.mdp]: why 3?

File: libgpdbcost/src/CCostModelGPDB.cpp (hunk 3 above)
Comment (d) [inline libgpdbcost/src/CCostModelGPDB.cpp]: 0. The 10% discount here is a good start, but I don't feel comfortable with it.
0. Logically, we know that this won't cause real motion, only the overhead of hashing, does the numbers from `CCostModelParamsGPDB::EcpHJHashTableColumnCostUnit` and `CCostModelParamsGPDB::EcpHJHashTableWidthCostUnit` make sense to you? If yes, how would you like to add a separate case to have clearer intent?

File: libgpopt/include/gpopt/base/CDistributionSpecHashedNoOp.h (hunk 1 above)
Comment (xinzweb) [inline libgpopt/include/gpopt/base/CDistributionSpecHashedNoOp.h]: Sorry, it should actually be `HASHED NO-OP`.

File: libgpopt/include/gpopt/base/CDistributionSpecStrictHashed.h [hunk 8]
```diff
@@ -18,6 +18,11 @@ namespace gpopt
 
 			virtual EDistributionType Edt() const;
 
+			virtual
+			const CHAR *SzId() const
+			{
+				return "STRICT HASH";
```
Comment (xinzweb) [inline libgpopt/include/gpopt/base/CDistributionSpecStrictHashed.h]: Sorry, it should actually be `STRICT HASHED`

File: libgpopt/include/gpopt/base/CDistributionSpecHashed.h [hunk 9]
```diff
@@ -88,7 +88,13 @@ namespace gpopt
 			{
 				return CDistributionSpec::EdtHashed;
 			}
-			
+
+			virtual
+			const CHAR *SzId() const
+			{
+				return "HASH";
```
Comment (xinzweb) [inline libgpopt/include/gpopt/base/CDistributionSpecHashed.h]: This should be `HASHED`

File: data/dxl/minidump/CPhysicalParallelUnionAllTest/ParallelUnion-ConstTable.mdp (hunk 7 above)
Comment (dhanashreek89) [inline data/dxl/minidump/CPhysicalParallelUnionAllTest/ParallelUnion-ConstTable.mdp]: Same as above.

**Comment by xinzweb** [comment]
@d please take a look at the new fix. Thanks.

File: data/dxl/minidump/CPhysicalParallelUnionAllTest/ParallelUnion-ConstTable.mdp (hunk 6 above)
Comment (vraghavan78) [inline data/dxl/minidump/CPhysicalParallelUnionAllTest/ParallelUnion-ConstTable.mdp]: why did we pick the plan id? can you find out?

File: data/dxl/minidump/CPhysicalParallelUnionAllTest/ParallelUnionAllWithNoRedistributableColumns.mdp [hunk 10]
```diff
@@ -1,8 +1,9 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <!--
-CREATE TABLE tbl1 (a int);
-CREATE TABLE tbl2 (b int);
+CREATE TABLE tbl1 (a int) DISTRIBUTE BY(a);
+CREATE TABLE tbl2 (b int) DISTRIBUTE BY(b);
 
+/*We expect the HASHED NO-OP*/
 EXPLAIN SELECT xmin FROM tbl1 UNION ALL SELECT xmin FROM tbl2;
```
Comment (vraghavan78) [inline data/dxl/minidump/CPhysicalParallelUnionAllTest/ParallelUnionAllWithNoRedistributableColumns.mdp]: can we have a query that does UNION also?

**Comment by dhanashreek89** [comment]
@vraghavan78 @d Please take a look. Thanks.

**Pull request merged**